docs/design/datacontracts/RuntimeTypeSystem.md

```diff
@@ -92,12 +92,49 @@ struct MethodDescHandle
 }
 ```
 
+```csharp
+public enum ArrayFunctionType
+{
+    Get = 0,
+    Set = 1,
+    Address = 2,
+    Constructor = 3
+}
+```
+
 ```csharp
 partial interface IRuntimeTypeSystem : IContract
 {
     public virtual MethodDescHandle GetMethodDescHandle(TargetPointer methodDescPointer);
 
     public virtual TargetPointer GetMethodTable(MethodDescHandle methodDesc);
+
+    // Return true for an uninstantiated generic method
+    public virtual bool IsGenericMethodDefinition(MethodDescHandle methodDesc);
+
+    public virtual ReadOnlySpan<TypeHandle> GetGenericMethodInstantiation(MethodDescHandle methodDesc);
+
+    // Return mdTokenNil (0x06000000) if the method doesn't have a token, otherwise return the token of the method
+    public virtual uint GetMethodToken(MethodDescHandle methodDesc);
+
+    // Return true if a MethodDesc represents an array method
+    // An array method is also a StoredSigMethodDesc
+    public virtual bool IsArrayMethod(MethodDescHandle methodDesc, out ArrayFunctionType functionType);
+
+    // Return true if a MethodDesc represents a dynamically generated method, either an IL Stub dynamically
+    // generated by the runtime, or a MethodDesc that describes a method represented by the System.Reflection.Emit.DynamicMethod class
+    // A dynamic method is also a StoredSigMethodDesc
+    public virtual bool IsDynamicMethod(MethodDescHandle methodDesc, out ReadOnlySpan<byte> methodName);
+
+    // A StoredSigMethodDesc is a MethodDesc for which the signature isn't found in metadata.
+    public virtual bool IsStoredSigMethodDesc(MethodDescHandle methodDesc, out ReadOnlySpan<byte> signature);
+
+    // Return true for a MethodDesc that describes a method represented by the System.Reflection.Emit.DynamicMethod class
+    // A LCG method is also a StoredSigMethodDesc
+    public virtual bool IsLCGMethod(MethodDescHandle methodDesc);
```

Comment:
`DynamicMethod` is the public name of the LCG feature.

ReflectionEmit is more general. Some methods emitted using Reflection.Emit are dynamic methods, some are not.

Reply:
... whereas I think of Reflection.Emit as a thing that only applies to things that generate the full metadata and types which is also not completely accurate...   The actual term in the BCL is that an LCGMethod maps to a method represented by the `System.Reflection.Emit.DynamicMethod` class, but calling it a DynamicMethod... might be extra special confusing, since the runtime has multiple meanings of what a `DynamicMethodDesc` represents, only 1 of which is a `System.Reflection.Emit.DynamicMethod`. (And of course LCG stands for Lightweight CodeGen, which is an internal codename that probably doesn't still need to be kept around.) I'd like to hear what @lambdageek, and @AaronRobinsonMSFT  think on this, as I know I'm not objective here as I've worked on this stuff for FAR too long to see things reasonably. Possibly the right approach would be to call it `DynamicMethod` and sometime in .NET 10, rename the `DynamicMethodDesc` to something like `RuntimeILMethodDesc`.

Reply:
I like the rename that elinor mentioned above at https://github.com/dotnet/runtime/pull/106169#discussion_r1712449039.

> rename the `DynamicMethodDesc` to something like `RuntimeILMethodDesc`.

I also agree with your suggestion above. I could also see `GeneratedMethodDesc` or `ILEmitMethodDesc`.

Reply:
I like `IsLCGMethod` -> `IsDynamicMethod` and `DynamicMethodDesc` ->`ILEmitMethodDesc`.

Not sure about `IsDynamicMethod` -> `IsDynamicallyGeneratedMethod`. Would just `IsGeneratedILMethod` be too broad?

Reply:
> DynamicMethodDesc ->ILEmitMethodDesc.

How about `DynamicMethodDesc` ->`NoMetadataMethodDesc` and `IsDynamicMethod` -> `IsNoMetadata` (existing property)?

Part of the codebase uses the NoMetadata name already: 
https://github.com/dotnet/runtime/blob/fa33eddf3533dac554e959da92f54104c0568c1e/src/coreclr/vm/method.hpp#L626-L630 . Notice that `MethodDesc::IsDynamicMethod` and `MethodDesc::IsNoMetadata` have the same implementation.

The key property of the `DynamicMethodDesc` is that it has no ECMA-335 metadata, it has no metadata token, etc.

Dynamic IL generation is not the key property of `DynamicMethodDesc`. The regular ECMA-335 MethodDescs can have dynamically generated (IL) code too. For example, `UnsafeAccessors` have dynamically generated IL but they are regular MethodDescs.

Reply:
I like @jkotas's idea here. Then the IsDynamicMethod api at the DAC level makes total sense, the existing `DynamicMethodDesc` translates to `NoMetadataMethodDesc` and gets a better name that more matches its utility, and we get rid of having both `MethodDesc::IsNoMetadata` and `MethodDesc::IsDynamicMethod` which are today exactly the same thing. While I'm at it, I'll rename `mcDynamic` to `mcNoMetadata`

Reply:
I'm trending toward the idea of putting this rename in a separate PR. Opinions?

Reply:
Offline discussion is to put together a separate PR for that. I'll just add a commit to update the cdac contract side naming to IsDynamicMethod, and the renaming of the underlying structures to be less confusing will be a separate PR that will probably miss .NET 9.